synapse/app/_daemonize.py

```diff
@@ -0,0 +1,156 @@
+# -*- coding: utf-8 -*
+# Inspired by https://github.com/thesharp/daemonize/blob/master/daemonize.py
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import fcntl
+import os
+import sys
+import signal
+import atexit
+import traceback
+
+
+class Daemonize(object):
+
+    """
+    Daemonize object.
+
+    Object constructor expects three arguments.
+
+    :param app: contains the application name which will be sent to syslog.
+    :param pid: path to the pidfile.
+    :param action: your custom function which will be executed after daemonization.
+    :param logger: use this logger object instead of creating new one, if provided.
+    :param chdir: change working directory if provided or /
+    :param outfile: File to redirect stdio to
```

Comment:
I'd prefer it if we could avoid changing the functionality in the same PR as the refactor, particularly since we're hardcoding `homeserver.log` here.

For now I suggest replicating the existing behaviour (ie, removing the `outfile` and `void_stdio` params, and continuing to redirect to /dev/null). We can introduce an alternative output file in a later PR.